File: android/app/src/main/java/org/openbot/tflite/Detector.java
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

// Modified by Matthias Mueller - Intel Intelligent Systems Lab - 2020

package org.openbot.tflite;

import android.app.Activity;
import android.graphics.Bitmap;
import android.graphics.RectF;
import android.os.SystemClock;
import android.os.Trace;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.PriorityQueue;

/**
 * Wrapper for frozen detection models trained using the Tensorflow Object Detection API:
 * github.com/tensorflow/models/tree/master/research/object_detection
 */
public abstract class Detector extends Network {

  /** Labels corresponding to the output of the vision model. */
  protected List<String> labels;

  // Number of output detections
  protected int NUM_DETECTIONS;

  protected String modelPath;
  protected int imageSizeX;
  protected int imageSizeY;

  /**
   * Creates a detector with the provided configuration.
   *
   * @param activity The current Activity.
   * @param model The model to use for classification.
   * @param device The device to use for classification.
   * @param numThreads The number of threads to use for classification.
   * @return A detector with the desired configuration.
   */
  public static Detector create(Activity activity, Model model, Device device, int numThreads)
      throws IOException {
    switch (model.classType) {
      case MOBILENETV1_1_0_Q:
      case MOBILENETV3_S_Q:
        return new DetectorQuantizedMobileNet(activity, model, device, numThreads);
      case YOLOV4:
        return new DetectorFloatYoloV4(activity, model, device, numThreads);
      default:
        return null;
    }
  }

  /** An immutable result returned by a Classifier/Detector describing what was recognized. */
  public static class Recognition {
    /**
     * A unique identifier for what has been recognized. Specific to the class, not the instance of
     * the object.
     */
    private final String id;

    /** Display name for the recognition. */
    private final String title;

    /**
     * A sortable score for how good the recognition is relative to others. Higher should be better.
     */
    private final Float confidence;

    /** Location within the source image for the location of the recognized object. */
    private RectF location;

    /** Detected class of the recognized object. */
    private int classId;

    public Recognition(
        final String id,
        final String title,
        final Float confidence,
        final RectF location,
        int classId) {
      this.id = id;
      this.title = title;
      this.confidence = confidence;
      this.location = location;
      this.classId = classId;
    }

    public String getId() {
      return id;
    }

    public String getTitle() {
      return title;
    }

    public Float getConfidence() {
      return confidence;
    }

    public RectF getLocation() {
      return new RectF(location);
    }

    public void setLocation(RectF location) {
      this.location = location;
    }

    public int getClassId() {
      return classId;
    }

    @Override
    public String toString() {
      String resultString = "";
      if (id != null) {
        resultString += "[" + id + "] ";
      }

      if (title != null) {
        resultString += title + " ";
      }

      if (confidence != null) {
        resultString += String.format(Locale.US, "(%.1f%%) ", confidence * 100.0f);
      }

      if (location != null) {
        resultString += location + " ";
      }

      return resultString.trim();
    }
  }

  /** Initializes a {@code Detector}. */
  protected Detector(Activity activity, Model model, Device device, int numThreads)
      throws IOException {
    super(activity, model, device, numThreads);

    labels = loadLabelList(activity);
    parseTflite();
    LOGGER.d("Created a Tensorflow Lite Detector.");
  }

  /** Reads label list from Assets. */
  private List<String> loadLabelList(Activity activity) throws IOException {
    List<String> labels = new ArrayList<String>();
    BufferedReader reader =
        new BufferedReader(new InputStreamReader(activity.getAssets().open(getLabelPath())));
    String line;
    while ((line = reader.readLine()) != null) {
      labels.add(line);
    }
    reader.close();
    return labels;
  }

  public List<Recognition> recognizeImage(final Bitmap bitmap, String className)
      throws IllegalArgumentException {
    // Log this method so that it can be analyzed with systrace.
    Trace.beginSection("recognizeImage");

    Trace.beginSection("preprocessBitmap");
    convertBitmapToByteBuffer(bitmap);
    Trace.endSection(); // preprocessBitmap

    // Copy the input data into TensorFlow.
    Trace.beginSection("feed");
    feedData();
    Trace.endSection();

    // Run the inference call.
    Trace.beginSection("runInference");
    long startTime = SystemClock.elapsedRealtime();
    runInference();
    long endTime = SystemClock.elapsedRealtime();
    Trace.endSection();
    LOGGER.v("Timecost to run model inference: " + (endTime - startTime));

    Trace.endSection(); // "recognizeImage"
    return getRecognitions(className);
  }

  protected float mNmsThresh = 0.25f;

  // non maximum suppression
  protected ArrayList<Recognition> nms(ArrayList<Recognition> list) {
    ArrayList<Recognition> nmsList = new ArrayList<Recognition>();

    for (int k = 0; k < labels.size(); k++) {
      // 1. Find max confidence per class
      PriorityQueue<Recognition> pq =
          new PriorityQueue<Recognition>(
              50,
              new Comparator<Recognition>() {
                @Override
                public int compare(final Recognition lhs, final Recognition rhs) {
                  // Intentionally reversed to put high confidence at the head of the queue.
                  return Float.compare(rhs.getConfidence(), lhs.getConfidence());
                }
              });

      for (int i = 0; i < list.size(); ++i) {
        if (list.get(i).getClassId() == k) {
          pq.add(list.get(i));
        }
      }

      // 2. Do non maximum suppression
      while (pq.size() > 0) {
        // insert detection with max confidence
        Recognition[] a = new Recognition[pq.size()];
        Recognition[] detections = pq.toArray(a);
        Recognition max = detections[0];
        nmsList.add(max);
        pq.clear();

        for (int j = 1; j < detections.length; j++) {
          Recognition detection = detections[j];
          RectF b = detection.getLocation();
          if (box_iou(max.getLocation(), b) < mNmsThresh) {
            pq.add(detection);
          }
        }
      }
    }
    return nmsList;
  }

  protected float box_iou(RectF a, RectF b) {
    return box_intersection(a, b) / box_union(a, b);
  }

  protected float box_intersection(RectF a, RectF b) {
    float w =
        overlap((a.left + a.right) / 2, a.right - a.left, (b.left + b.right) / 2, b.right - b.left);
    float h =
        overlap((a.top + a.bottom) / 2, a.bottom - a.top, (b.top + b.bottom) / 2, b.bottom - b.top);
    if (w < 0 || h < 0) return 0;
    float area = w * h;
    return area;
  }

  protected float box_union(RectF a, RectF b) {
    float i = box_intersection(a, b);
    float u = (a.right - a.left) * (a.bottom - a.top) + (b.right - b.left) * (b.bottom - b.top) - i;
    return u;
  }

  protected float overlap(float x1, float w1, float x2, float w2) {
    float l1 = x1 - w1 / 2;
    float l2 = x2 - w2 / 2;
    float left = l1 > l2 ? l1 : l2;
    float r1 = x1 + w1 / 2;
    float r2 = x2 + w2 / 2;
    float right = r1 < r2 ? r1 : r2;
    return right - left;
  }

  /**
   * Get the name of the label file stored in Assets.
   *
   * @return
   */
  protected abstract String getLabelPath();

  /**
   * Run inference using the prepared input in {@link #imgData}. Afterwards, the result will be
   * provided by getProbability().
   *
   * <p>This additional method is necessary, because we don't have a common base for different
   * primitive data types.
   */
  protected abstract void runInference();

  /**
   * Feeds the data
   *
   * <p>This additional method is necessary, because we can have different number of detections
   */
  protected abstract void feedData();

  /**
   * Get the total number of labels.
   *
   * @return
   */
  protected int getNumLabels() {
    return labels.size();
  }

  public List<String> getLabels() {
    List<String> list = new ArrayList<>();
    for (String label : labels) if (!label.equals("???")) list.add(label);
    return list;
  }
  /**
   * Get the number of detections.
   *
   * @return
   */
  protected abstract int getNumDetections();

  /** Get specs from tflite file */
  protected abstract void parseTflite();

  /**
   * Get the recognitions.
   *
   * @return
   */
  protected abstract List<Recognition> getRecognitions(String className);
}

File: ios/OpenBot/OpenBot/tflite/Detector.swift
//
// Created by Sparsh Jain on 01/11/22.
//

import Foundation
import UIKit

/// Wrapper for frozen detection models trained using the Tensorflow Object Detection API:
/// github.com/tensorflow/models/tree/master/research/object_detection
class Detector: Network {
    var labels: [String] = []; // Labels corresponding to the output of the vision model.
    var NUM_DETECTIONS: Int = 0; // Number of output detections
    var selectedClass: String?;
    var selectedMultipleClass: [String] = [];
    let ciContext = CIContext()
    let imageSizeX: Int = -1;
    let imageSizeY: Int = -1;

    /// Creates a detector with the provided configuration.
    ///
    /// - Parameters
    ///   - model: The model to use for classification.
    ///   - device: The device to use for classification.
    ///   - numThreads: The number of threads to use for classification.
    /// - Returns: A detector with the desired configuration.
    ///
    static func create(model: Model, device: RuntimeDevice, numThreads: Int) throws -> AnyObject? {
        switch (model.classType) {
        case .MOBILENET, .EFFICIENTDET:
            return try DetectorDefault(model: model, device: device, numThreads: numThreads);
        case .YOLOV4:
            return try DetectorFloatYoloV4(model: model, device: device, numThreads: numThreads);
        case .YOLOV5:
            return try DetectorFloatYoloV5(model: model, device: device, numThreads: numThreads);
        default:
            return nil;
        }
    }

    /// An immutable result returned by a Classifier/Detector describing what was recognized.
    class Recognition {

        // A unique identifier for what has been recognized. Specific to the class, not the instance of the object.
        private var id: String;

        // Display name for the recognition.
        private var title: String;

        // A sortable score for how good the recognition is relative to others. Higher should be better.
        private var confidence: Float;

        // Location within the source image for the location of the recognized object.
        private var location: CGRect;

        // Detected class of the recognized object.
        private var classId: Int;

        /// Initialization routine
        ///
        /// - Parameters:
        ///     - id:
        ///     - title:
        ///     - confidence:
        ///     - location:
        ///     - classId:
        init(id: String, title: String, confidence: Float, location: CGRect, classId: Int) {
            self.id = id;
            self.title = title;
            self.confidence = confidence;
            self.location = location;
            self.classId = classId;
        }

        /// Compare two Recognition objects in terms of detection confidence
        ///
        /// - Parameters:
        ///     - lhs: first Recognition object
        ///     - rhs: second Recognition object
        static func >(lhs: Recognition, rhs: Recognition) -> Bool {
            lhs.confidence > rhs.confidence
        }

        /// Getter function.
        ///
        /// - Returns: id of the recognition
        public func getId() -> String {
            id;
        }

        /// Getter function.
        ///
        /// - Returns: class Id
        public func getClassId() -> Int {
            classId;
        }

        /// Getter function.
        ///
        /// - Returns: name of the recognition
        public func getTitle() -> String {
            title;
        }

        /// Getter function.
        ///
        /// - Returns: confidence of the recognition
        public func getConfidence() -> Float {
            confidence;
        }

        /// Getter function.
        ///
        /// - Returns: bounding box of the recognition
        public func getLocation() -> CGRect {
            location;
        }

        /// Setter function.
        ///
        /// - Parameters: bounding box of the recognition
        public func setLocation(location: CGRect) {
            self.location = location;
        }
    }

    /// Initializes a Detector.
    ///
    /// - Parameters:
    ///     - model: the model considered in the inference process
    ///     - device: CPU, GPU or XNNPACK (neural engine)
    ///     - numThreads: number of threads used tin the inference process
    override init(model: Model, device: RuntimeDevice, numThreads: Int) throws {
        try super.init(model: model, device: device, numThreads: numThreads);
        NotificationCenter.default.addObserver(self, selector: #selector(updateObject), name: .updateObject, object: nil)
        labels = loadLabelList(filePath: getLabelPath());
        selectedClass = labels.first {
            $0.capitalized == "person".capitalized
        };
        selectedMultipleClass = ["person"]
        parseTFlite();
    }

    /// Reads label list from Assets.
    ///
    /// - Parameters: path of the text file containig the labels.
    /// - Returns: list of labels contained in the file.
    func loadLabelList(filePath: String) -> [String] {
        var result: [String] = []
        if let filepath = Bundle.main.path(forResource: filePath, ofType: "") {
            do {
                let contents = try String(contentsOfFile: filepath)
                let output = contents.components(separatedBy: "\n")
                for label in output {
                    if (label != " " && label != "") {
                        result.append(label);
                    }
                }
            } catch {
                print("cannot convert file content to string");
            }
        } else {
            print("labelmap.txt or coco.txt not found");
        }
        return result
    }

    /// Propagate an image through a neural network to recognize objects of a predefined class.
    ///
    /// - Parameters: The buffer containing the image.
    /// - Returns: A list of the recognized objects for a given label
    func recognizeImage(pixelBuffer: CVPixelBuffer, detectionType: String) -> [Recognition] {

        let imageWidth = CVPixelBufferGetWidth(pixelBuffer)
        let imageHeight = CVPixelBufferGetHeight(pixelBuffer)
        let sourcePixelFormat = CVPixelBufferGetPixelFormatType(pixelBuffer)
        assert(sourcePixelFormat == kCVPixelFormatType_32ARGB ||
                sourcePixelFormat == kCVPixelFormatType_32BGRA ||
                sourcePixelFormat == kCVPixelFormatType_32RGBA)

        // Crops the image to the biggest square in the center and scales it down to model dimensions.
        let scaledSize = CGSize(width: getImageSizeX(), height: getImageSizeY())
        guard let scaledPixelBuffer = pixelBuffer.resized(to: scaledSize, preserveAspectRatio: false, with: self.preAllocatedMemoryPool!) else {
            return []
        }

        do {

            // Pre-process input image (orientation, scaling, cropping...).
            let inputTensor = try tflite!.input(at: 0);
            guard let rgbData = rgbDataFromBuffer(scaledPixelBuffer, isModelQuantized: inputTensor.dataType == .uInt8) else {
                return []
            }

            // Copy the input data into TensorFlow.
            try tflite?.copy(rgbData, toInputAt: 0);
            try feedData();

            // Run the inference call.
            try tflite?.invoke();

            if (detectionType == "multiple") {
                return getMultipleRecognitions(classA: selectedMultipleClass[0], classB: selectedMultipleClass[1], width: imageWidth, height: imageHeight)
            } else if (detectionType == "all") {
                return getAllRecognitions(width: imageWidth, height: imageHeight)
            } else {
                return getRecognitions(className: selectedClass!, width: imageWidth, height: imageHeight);
            }
            // Post-Processing.

        } catch {
            print("error:\(error)")
            if (detectionType == "multiple") {
                return getMultipleRecognitions(classA: selectedMultipleClass[0], classB: selectedMultipleClass[1], width: imageWidth, height: imageHeight)
            } else if (detectionType == "all") {
                return getAllRecognitions(width: imageWidth, height: imageHeight)
            } else {
                return getRecognitions(className: selectedClass!, width: imageWidth, height: imageHeight);
            }
        };
    }

    var mNmsThresh: Float = 0.25;

    func getObjThresh() -> Float {
        // TODO: return ObjectTrackingFragment.MINIMUM_CONFIDENCE_TF_OD_API
        0.5;
    }

    /// Non maximum suppression
    ///
    /// - Parameters: An array of "Recognition" objects, built from the output of the neural detector.
    /// - Returns: Filtered array of "Recognition" objects.
    func nms(recognitions: [Recognition]) -> [Recognition] {

        var nmsList: [Recognition] = [];
        var pq = recognitions;
        pq.sort(by: >);
        while (pq.count > 0) {
            var detections = pq;
            detections.sort(by: >);
            let max = detections[0];
            nmsList.append(max);
            pq.removeAll();
            if (detections.count > 0) {
                let size = detections.count - 1;
                for i in stride(from: 1, to: size, by: 1) {
                    let detection: Recognition = detections[i];
                    let b = detection.getLocation();
                    if (box_iou(a: max.getLocation(), b: b) < mNmsThresh) {
                        pq.append(detection);
                    }
                }
            }
        }
        return nmsList;
    }

    /// Intersection over union.
    ///
    /// - Parameters:
    ///     - a: first bounding box.
    ///     - b: second bounding box.
    /// - Returns: IoU of bounding boxes a and b.
    func box_iou(a: CGRect, b: CGRect) -> Float {
        box_intersection(a: a, b: b) / box_union(a: a, b: b);
    }

    /// Intersection.
    ///
    /// - Parameters:
    ///     - a: first bounding box.
    ///     - b: second bounding box.
    /// - Returns: Intersection surface of bounding boxes a and b.
    func box_intersection(a: CGRect, b: CGRect) -> Float {
        let w = overlap(x1: Float(a.midX), w1: Float(a.width), x2: Float(b.midX), w2: Float(b.width));
        let h = overlap(x1: Float(a.midY), w1: Float(a.height), x2: Float(b.midY), w2: Float(b.height));
        if (w < 0 || h < 0) {
            return 0
        };
        return w * h;
    }

    /// Union.
    ///
    /// - Parameters:
    ///     - a: first bounding box.
    ///     - b: second bounding box.
    /// - Returns: Union surface of bounding boxes a and b.
    func box_union(a: CGRect, b: CGRect) -> Float {
        let i = box_intersection(a: a, b: b);
        let u = Float(a.width * a.height) + Float(b.width * b.height) - i;
        return u;
    }

    /// Segment overlap computation.
    ///
    /// - Parameters:
    ///     - x1: origin of the first segment.
    ///     - w1: length of the first segment.
    ///     - x2: origin of the second segment.
    ///     - w2: length of the second segment.
    /// - Returns: Overlap of the two segments.
    func overlap(x1: Float, w1: Float, x2: Float, w2: Float) -> Float {
        let l1: Float = x1 - w1 / 2;
        let l2: Float = x2 - w2 / 2;
        let left: Float = l1 > l2 ? l1 : l2;
        let r1 = x1 + w1 / 2;
        let r2 = x2 + w2 / 2;
        let right = r1 < r2 ? r1 : r2;
        return right - left;
    }

    /// Get the name of the label file stored in Assets.
    func getLabelPath() -> String {
        "";
    }

    /// Feeds the data.
    /// This additional method is necessary, because we can have different number of detections
    func feedData() throws {
    }

    /// Get the total number of labels.
    func getNumLabels() -> Int {
        labels.count
    }

    /// Get the list of all the labels
    func getLabels() -> [String] {
        var result: [String] = []

        for label in labels {
            if (label != "???" && label != "" && label != " ") {
                result.append(label);
            }
        }
        return result
    }

    /// Get the number of detections.
    func getNumDetections() -> Int {
        NUM_DETECTIONS;
    }

    /// Get specs from tflite file.
    func parseTFlite() {
    }

    /// Get the recognitions.
    func getRecognitions(className: String, width: Int, height: Int) -> [Recognition] {
        [];
    }

    /// Get the multiple recognitions
    func getMultipleRecognitions(classA: String, classB: String, width: Int, height: Int) -> [Recognition] {
        [];
    }

    func getAllRecognitions(width: Int, height: Int) -> [Recognition] {
        [];
    }

    ///
    @objc func updateObject(_ notification: Notification) {
        selectedClass = notification.object as? String
    }

    /// Setter function
    public func setSelectedClass(newClass: String) {
        selectedClass = newClass;
    }

    public func setMultipleSelectedClass(newClasses: [String]) {
        selectedMultipleClass = newClasses;
    }

    public func returnMultipleSelectedClass() -> [String] {
        selectedMultipleClass
    }
}

